### PR TITLE
Feature/validation model specs

### PR DESCRIPTION
**Pull request by 2004-okari**
This pull request addresses the project requirements by adding validations to the User and Post models, as well as creating unit specs (unit tests) to ensure that the models' methods and validations function as expected.

**User Model Validations:**
1. Added validation to ensure that the `name` attribute is not blank.

2. Added validation to ensure that the `posts_counter` attribute is an integer greater than or equal to zero.

**Post Model Validations:**
1. Added validation to ensure that the `title` attribute is not blank.

2. Added validation to ensure that the `title` attribute does not exceed 250 characters.

3. Added validation to ensure that the `comments_counter` attribute is an integer greater than or equal to zero.

4. Added validation to ensure that the `likes_counter` attribute is an integer greater than or equal to zero.

**Unit Specs (Unit Tests):**
1. Created unit specs for all models' methods and validations to thoroughly test their functionality.

These changes ensure that the models adhere to the specified data validation rules, improving data quality and integrity. The unit specs validate that the models' methods and validations work correctly and that the application functions as expected.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    it 'is valid with a name' do
+      user = User.new(name: 'John Doe', posts_counter: 3)
+      expect(user).to be_valid
+    end
+
+    it 'is not valid with a negative posts_counter' do
+      user = User.new(name: 'Alice', posts_counter: -3)
+      expect(user).to_not be_valid
+    end
```
Comment (Lucas-Masaba) [inline spec/models/user_spec.rb:18]: - [ ] Please make sure to test all your methods in your User model for example the `recent_posts` method.

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    expect(Post.new(author: @author, title: 'Okari Nyandika' * 250, text: 'Hello World!', comments_counter: 'one',
+                    likes_counter: 0)).to_not be_valid
+  end
+
+  it 'is invalid if likes_counter is less than zero' do
+    expect(Post.new(author: @author, title: 'Okari Nyandika' * 250, text: 'Hello World!', comments_counter: 1,
+                    likes_counter: -5)).to_not be_valid
+  end
+
+  context '#recent_comments' do
```
Comment (Lucas-Masaba) [inline spec/models/post_spec.rb:33]: - [ ] Please make sure to test all your methods in your User model for example the `recent_comments`, `update_posts_counter` methods. They should be called as you write their tests

File: spec/models/like_spec.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    it 'is not valid without post' do
+      expect(Like.new(user: @user)).to_not be_valid
+    end
+
+    it 'is not valid without user' do
+      expect(Like.new(post: @post)).to_not be_valid
+    end
+  end
+
+  context '#update_likes_counter' do
```
Comment (Lucas-Masaba) [inline spec/models/like_spec.rb:27]: - [x] Please make sure to test all your methods in your User model for example the `update_likes_counter` method.

File: spec/models/comment_spec.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    it 'is not valid without post' do
+      expect(Comment.new(user: @user)).to_not be_valid
+    end
+
+    it 'is not valid without user' do
+      expect(Comment.new(post: @post)).to_not be_valid
+    end
+  end
+
+  context '#update_comments_counter' do
```
Comment (Lucas-Masaba) [inline spec/models/comment_spec.rb:27]: - [x] Please make sure to test all your methods in your User model for example the `update_comments_counter` method.

**Review by Lucas-Masaba — CHANGES_REQUESTED** [review]
# Hello @2004-okari  👋 , 

Good job so far! 👍 
There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## Highlights 🍾 
- Linter checks passing :heavy_check_mark: 
- All written tests passing :heavy_check_mark: 
- Validations setup correctly :gear: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Lucas-Masaba in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb (hunk 2 above)
Comment (2004-okari) [reply spec/models/post_spec.rb:33]: I have only one method in my User model so I really don’t understand which methods you are referring to, and the post model is full tested @Lucas-Masaba

File: spec/models/like_spec.rb (hunk 3 above)
Comment (2004-okari) [reply spec/models/like_spec.rb:27]: I don't have that method in my User model. @Lucas-Masaba

File: spec/models/comment_spec.rb (hunk 4 above)
Comment (2004-okari) [reply spec/models/comment_spec.rb:27]: I don't have such a method in my user model @Lucas-Masaba

File: spec/models/post_spec.rb (hunk 2 above)
Comment (Lucas-Masaba) [reply spec/models/post_spec.rb:33]: Hello @2004-okari , I meant to say the Post model. That was a typo on my part. Sorry about that. These are the methods seen in your post model:
image

File: spec/models/like_spec.rb (hunk 3 above)
Comment (Lucas-Masaba) [reply spec/models/like_spec.rb:27]: Hello @2004-okari , I meant to say the Like model. That was a typo on my part. Sorry about that. These are the methods seen in your Like model:
image

File: spec/models/comment_spec.rb (hunk 4 above)
Comment (Lucas-Masaba) [reply spec/models/comment_spec.rb:27]: Hello @2004-okari , I meant to say the Comment model. That was a typo on my part. Sorry about that. These are the methods seen in your Comment model:
image

**Comment by Lucas-Masaba** [comment]
Hello @2004-okari , please let me know if my corrections clarified your queries.

File: spec/models/post_spec.rb (hunk 2 above)
Comment (2004-okari) [reply spec/models/post_spec.rb:33]: Hello @Lucas-Masaba ,
the 
```
 context '#update_posts_counter' do
    before :all do
      10.times do
        Post.create(author: @author, title: 'Kathy', text: 'Text here....', comments_counter: 0,
                    likes_counter: 0)
      end
    end

    it 'keeps track of posts and equals 10' do
      expect(@author.posts_counter).to eq 21
    end
  end```.

The #update_post_counter is a syntax for calling a method in the describe or context command. The recent post is a private method hence it cannot be tested hence I added a count_recent_post method that calls the private method hence it is tested indirectly

File: spec/models/like_spec.rb (hunk 3 above)
Comment (2004-okari) [reply spec/models/like_spec.rb:27]: hello @Lucas-Masaba ,I do have tests for that method
```
context '#update_likes_counter' do
    before :all do
      5.times { Like.create(user: @user, post: @post) }
    end

    it 'keeps track of likes and equals 5' do
      expect(@post.likes_counter).to eq 5
    end
  end
```

File: spec/models/comment_spec.rb (hunk 4 above)
Comment (2004-okari) [reply spec/models/comment_spec.rb:27]: Hello @Lucas-Masaba ,  I also have tests for the mentioned method
```
ontext '#update_comments_counter' do
    before :all do
      5.times { |comment_i| Comment.create(user: @user, post: @post, text: (comment_i + 1).to_s) }
    end

    it 'keeps track of comments and equals 5' do
      expect(@post.comments_counter).to eq 5
    end
  end
```

**Comment by 2004-okari** [comment]
> Hello @2004-okari , please let me know if my corrections clarified your queries.

@Lucas-Masaba , not really since I am only missing tests for one method

**Comment by Lucas-Masaba** [comment]
@2004-okari , I read through your explanation and looked at your tests again and noticed the methods I requested changes for were private and you tested them through a public method like you mentioned. Please proceed and ask for another review to get this milestone approved. Thank you for the clear communication and sorry for the inconvenience.

**Comment by 2004-okari** [comment]
> @2004-okari , I read through your explanation and looked at your tests again and noticed the methods I requested changes for were private and you tested them through a public method like you mentioned. Please proceed and ask for another review to get this milestone approved. Thank you for the clear communication and sorry for the inconvenience.

@Lucas-Masaba , No worries, Thanks for the feedback!

**Comment by tugsem** [comment]
## STATUS: APPROVED 👍🎯  
Hello @2004-okari 👋 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights ⭐ 
✔️ correct use of gitflow
✔️ linters are passing
✔️all test are passing
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**